# E-Commerce Public Analysis Project
---

First, import all the the library that we used

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Data Wrangling Session
---

## _Gathering Data_

import `customers_dataset.csv` data into dataframe

In [2]:
customers_df = pd.read_csv("./data/customers_dataset.csv")
customers_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

import `geolocation_dataset.csv` data into dataframe

In [3]:
geolocation_df = pd.read_csv("./data/geolocation_dataset.csv")
geolocation_df.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

import `order_item_dataset.csv` data into dataframe

In [4]:
order_items_df = pd.read_csv("./data/order_items_dataset.csv")
order_items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

import `order_payments_dataset.csv` data into dataframe

In [5]:
order_payments_df = pd.read_csv("./data/order_payments_dataset.csv")
order_payments_df.head()


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

import `order_reviews_dataset.csv` data into dataframe

In [6]:
order_reviews_df = pd.read_csv("./data/order_reviews_dataset.csv")
order_reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

import `orders_dataset.csv` data into dataframe

In [7]:
orders_df = pd.read_csv("./data/orders_dataset.csv")
orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

import `product_category_name_translation.csv` (pcn) data into dataframe

In [8]:
pcn_df = pd.read_csv("./data/product_category_name_translation.csv")
pcn_df.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

import `products_dataset.csv` data into dataframe

In [9]:
products_df = pd.read_csv("./data/products_dataset.csv")
products_df.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

import `sellers_dataset.csv` data into dataframe

In [10]:
sellers_df = pd.read_csv("./data/sellers_dataset.csv")
sellers_df.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

## _Assessing Data_

### Assessing to `customers_df`

In [11]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [12]:
print(f"\nNumber of duplications: {customers_df.duplicated().sum()}")
customers_df.describe(include='all')


Number of duplications: 0


customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        customer_zip_code_prefix customer_city customer_state  
count               99441.000000         99441          99441  
unique                       NaN          4119             27  
top                          NaN     sao paulo             SP  
freq                         NaN         15540          41746  
mean                35137.474583           NaN            NaN  
std                 29797.938996           NaN            NaN  
min                  1003.000000           NaN            NaN  
25%                 11347.000000           NaN            NaN  
50%                 24416.000000           NaN            NaN  
75%                 58900.000000           NaN            NaN  
max                 99990.000000           NaN            NaN

**interim analysis**: The dataset is already clean.

**note**: The zip code has already 5 digit, but leading zeros in integer literals are not permitted. so example if the zip code is 9780, the fact is 09780.

### Assessing to `geolocation_df`

In [13]:
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [14]:
print(f"\nNumber of duplications: {geolocation_df.duplicated().sum()}")
geolocation_df.describe()


Number of duplications: 261831


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02

**interim analysis**: There's 261831 duplicate data.

**note**: The zip code has already 5 digit, but leading zeros in integer literals are not permitted. so example if the zip code is 9780, the fact is 09780.

### Assessing to `order_items_df`

In [15]:
order_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [16]:
print(f"\nNumber of duplications: {order_items_df.duplicated().sum()}")
order_items_df.describe()


Number of duplications: 0


order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

**interim analysis**: There's a data type error for shipping_limit_date column.

### Assessing to `order_payments_df`

In [17]:
order_payments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [18]:
print(f"\nNumber of duplications: {order_payments_df.duplicated().sum()}")
order_payments_df.describe()


Number of duplications: 0


payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

**interim analysis**: The dataset is already clean.

### Assessing to `order_reviews_df`

In [19]:
order_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [20]:
order_reviews_df.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [21]:
print(f"\nNumber of duplications: {order_reviews_df.duplicated().sum()}")
order_reviews_df.describe()


Number of duplications: 0


review_score
count  99224.000000
mean       4.086421
std        1.347579
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

**interim analysis**: There are missing values in both the review_comment_title and review_comment_message columns, with a total of 876565 and 58247 missing values, respectively. and there're a data type error for review_creation_date and review_answer_timestamp column.

### Assessing to `orders_df`

In [22]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [23]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [24]:
print(f"\nNumber of duplications: {orders_df.duplicated().sum()}")
orders_df.describe()


Number of duplications: 0


order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

**interim analysis**: There are missing values in the order_approved_at, order_delivered_carrier_date, and order_delivered_customer_date, with a total of 160, 1783 and 2965 missing values, respectively. and there're data type error for order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date column.

### Assessing to `pcn_df`

In [25]:
pcn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [26]:
print(f"\nNumber of duplications: {pcn_df.duplicated().sum()}")
pcn_df.describe()


Number of duplications: 0


product_category_name product_category_name_english
count                     71                            71
unique                    71                            71
top             beleza_saude                 health_beauty
freq                       1                             1

**interim analysis**: dataset is already clean.

### Assessing to `products_df`

In [27]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [28]:
products_df.isna().sum()


product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [29]:
print(f"\nNumber of duplications: {products_df.duplicated().sum()}")
products_df.describe()


Number of duplications: 0


product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

**interim analysis**: dataset column except product_id there are missing value.

### Assessing to `sellers_df`

In [30]:
sellers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [31]:
print(f"\nNumber of duplications: {sellers_df.duplicated().sum()}")
sellers_df.describe()


Number of duplications: 0


seller_zip_code_prefix
count             3095.000000
mean             32291.059451
std              32713.453830
min               1001.000000
25%               7093.500000
50%              14940.000000
75%              64552.500000
max              99730.000000

**interim analysis**: The dataset is already clean.

**note**: The zip code has already 5 digit, but leading zeros in integer literals are not permitted. so example if the zip code is 9780, the fact is 09780.

### Summary of Data Assessment
---

#### `customers_df`
- The dataset is already clean.

#### `geolocation_df`
- There are 261,831 duplicate records.

#### `order_items_df`
- There's a data type error for shipping_limit_date column.

#### `order_payments_df`
- The dataset is already clean.

#### `order_reviews_df`
- Missing Data: There are missing values in both the `review_comment_title` and `review_comment_message` columns, with a total of 876,565 and 58,247 missing values, respectively.
- There're a data type error for review_creation_date and review_answer_timestamp column.

#### `orders_df`
- Missing Data: There are missing values in the `order_approved_at`, `order_delivered_carrier_date`, and `order_delivered_customer_date` columns, with a total of 160, 1,783, and 2,965 missing values, respectively.
- There're data type error for order_delivered_carrier_date, order_delivered_customer_date, order_estimated_delivery_date column.

#### `pcn_df`
- The dataset is already clean.

#### `products_df`
- Missing Values: Except for the `product_id` column, there are missing values in other columns.

#### `sellers_df`
- The dataset is already clean.


## _Cleaning Data_

### Cleaning data `geolocation_df`

drop the duplicates data

In [32]:
geolocation_df.drop_duplicates(inplace=True)

In [33]:
print(f"Number of duplications: {geolocation_df.duplicated().sum()}")

Number of duplications: 0


### Cleaning data `order_items_df`

convert "shipping_limit_date" data type from object to datetime

In [34]:
order_items_df["shipping_limit_date"] = pd.to_datetime(
    order_items_df["shipping_limit_date"])


In [35]:
order_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  object        
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  object        
 3   seller_id            112650 non-null  object        
 4   shipping_limit_date  112650 non-null  datetime64[ns]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(3)
memory usage: 6.0+ MB


### Cleaning data `order_reviews_df`

#### Fill NA Review Comments Title

In [36]:
order_reviews_df.review_comment_title.isna().sum()

87656

for objective reason, so the missing value will be place to "Prefer not to say".

In [37]:
order_reviews_df.review_comment_title.fillna(
  value="Prefer not to say",
  inplace=True)

In [38]:
order_reviews_df.review_comment_title.isna().sum()


0

In [39]:
order_reviews_df.review_comment_title.value_counts()

review_comment_title
Prefer not to say            87656
Recomendo                      423
recomendo                      345
Bom                            293
super recomendo                270
                             ...  
Luminaria sobrepor led 18        1
Não recebi.                      1
produto com lacre violado        1
Atraso Entrega                   1
Foto enganosa                    1
Name: count, Length: 4528, dtype: int64

#### Fill NA Review Comments Message

In [40]:
order_reviews_df.review_comment_message.isna().sum()

58247

In [41]:
order_reviews_df.review_comment_message.fillna(
    value="Prefer not to say",
    inplace=True)


In [42]:
order_reviews_df.review_comment_message.value_counts()


review_comment_message
Prefer not to say                                                                                                                                                                                      58247
Muito bom                                                                                                                                                                                                230
Bom                                                                                                                                                                                                      189
muito bom                                                                                                                                                                                                122
bom                                                                                                                                                          

In [43]:
order_reviews_df.isna().sum()

review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation_date       0
review_answer_timestamp    0
dtype: int64

#### Fix Data Type Error

In [44]:
datetime_column = ["review_creation_date", "review_answer_timestamp"]

for column in datetime_column:
  order_reviews_df[column] = pd.to_datetime(order_reviews_df[column])


In [45]:
order_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  object        
 1   order_id                 99224 non-null  object        
 2   review_score             99224 non-null  int64         
 3   review_comment_title     99224 non-null  object        
 4   review_comment_message   99224 non-null  object        
 5   review_creation_date     99224 non-null  datetime64[ns]
 6   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB


### Cleaning data `orders_df`

#### Fix Data Type Error

In [46]:
order_datetime_column = ["order_purchase_timestamp",
                         "order_approved_at",
                         "order_delivered_carrier_date",
                         "order_delivered_customer_date",
                         "order_estimated_delivery_date"]

for column in order_datetime_column:
  orders_df[column] = pd.to_datetime(orders_df[column])

In [47]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_approved_at              99281 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  96476 non-null  datetime64[ns]
 7   order_estimated_delivery_date  99441 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.1+ MB


#### Fix Missing Value

First, we drop "order_delivered_carrier_date" column that have missing value. Because, if we fill with something, we can't make sure it's objective and based on fact.

In [48]:
orders_df.dropna(subset=['order_delivered_carrier_date'], inplace=True)

In [49]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  14
order_delivered_carrier_date        0
order_delivered_customer_date    1183
order_estimated_delivery_date       0
dtype: int64

Secondly, we can fill "order_approved_at" with the delta of user purchase timestamp and approved duration mean

In [50]:
orders_df['order_approved_at'] = pd.to_datetime(orders_df['order_approved_at'])
approved_rate = orders_df['order_approved_at'] - orders_df['order_purchase_timestamp']
approved_rate = approved_rate.apply(lambda x: x.total_seconds())
orders_df['approved_rate'] = round(approved_rate/60)
orders_df.sample(5)

order_id                       customer_id  \
48313  37f6e52783daf8d5b0a61c8876c23b38  05d4a0e7365cf69b45d3af76236ca83a   
87506  91c3ea3f720dd9cc12b762347379b0eb  930cddd48e6ae47a5922cff200d54069   
11215  5f5e8d13d1088bb941ae3b0ebff0feff  a775583c19646a68288f58bdfb1b90c7   
92173  8a7aac96934b61b070e94762ca857e31  29a61d310ca39b308cd0cd0115f4259b   
55478  2f40163a18439bbc77fa0f1984bf2596  57e873ebe362278f24712569d2d32ce6   

      order_status order_purchase_timestamp   order_approved_at  \
48313    delivered      2018-02-19 12:08:24 2018-02-19 12:27:17   
87506    delivered      2017-08-25 21:13:15 2017-08-26 02:10:11   
11215    delivered      2017-05-12 20:52:38 2017-05-13 02:50:27   
92173    delivered      2018-01-23 08:44:12 2018-01-25 02:57:43   
55478    delivered      2017-12-04 12:07:53 2017-12-04 16:33:21   

      order_delivered_carrier_date order_delivered_customer_date  \
48313          2018-02-21 10:52:03           2018-03-09 12:48:28   
87506          2017-08-28 20:33:52           2017-09-05 17:12:03   
11215          2017-05-16 09:59:53           2017-05-18 16:34:07   
92173          2018-01-26 00:23:19           2018-02-01 22:25:23   
55478          2017-12-08 10:03:54           2017-12-22 20:52:17   

      order_estimated_delivery_date  approved_rate  
48313                    2018-03-16           19.0  
87506                    2017-09-14          297.0  
11215                    2017-05-25          358.0  
92173                    2018-02-21         2534.0  
55478                    2017-12-20          265.0

In [51]:
mean_approved_rate = orders_df["approved_rate"].mean()
print(f"Approved Rate Mean: {mean_approved_rate} minutes")

orders_df['order_approved_at'].fillna(
    value=orders_df['order_purchase_timestamp'] +
    pd.to_timedelta(mean_approved_rate, unit='m'),
    inplace=True
)

orders_df.drop("approved_rate", axis=1, inplace=True)

Approved Rate Mean: 617.4651693908484 minutes


In [52]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date        0
order_delivered_customer_date    1183
order_estimated_delivery_date       0
dtype: int64

same as before, we can fill "order_delivered_customer_date" with the delta of order_estimated_delivery_date minus factual delivered duration mean

In [54]:
fact_delivery_dur = orders_df["order_estimated_delivery_date"] - orders_df["order_delivered_customer_date"]
fact_delivery_dur = fact_delivery_dur.apply(lambda x: x.total_seconds())
orders_df["fact_delivery_dur"] = round(fact_delivery_dur/86400)
orders_df.sample(5)

order_id                       customer_id  \
95232  52cfb42088bb202a56a77b48a3051baa  0031c1062edf836ef7a7a7ea758b137b   
20437  b886aff8576c2f1f035c507ba7fd6a9c  28b41c7b6f3d16d027bfea553f86c827   
79437  ac95b87f13bbd7609c75bfcc17cf3ae7  c1cc319cd4cc87a520c08d8e403bed0b   
42865  6113734f28c4dd1490de62945a9adf43  9c8f4989367caf5cc9fc08254f3ef547   
68172  68b73b2773ff90dc774d217f5a554680  62faee0a2a0c1b99966bdc94c12081d2   

      order_status order_purchase_timestamp   order_approved_at  \
95232    delivered      2017-03-01 20:00:23 2017-03-02 20:05:14   
20437    delivered      2018-04-20 16:29:38 2018-04-24 19:23:55   
79437    delivered      2018-07-20 14:50:31 2018-07-20 15:04:22   
42865    delivered      2017-11-29 22:59:41 2017-11-29 23:17:58   
68172    delivered      2018-01-08 15:28:25 2018-01-08 17:53:18   

      order_delivered_carrier_date order_delivered_customer_date  \
95232          2017-03-09 09:24:05           2017-03-13 12:08:59   
20437          2018-04-23 20:46:37           2018-05-03 21:45:47   
79437          2018-07-20 17:13:00           2018-07-27 17:04:21   
42865          2017-12-15 01:33:19           2017-12-28 19:39:05   
68172          2018-01-09 23:32:29           2018-01-21 13:09:58   

      order_estimated_delivery_date  fact_delivery_dur  
95232                    2017-03-20                6.0  
20437                    2018-05-23               19.0  
79437                    2018-08-01                4.0  
42865                    2018-01-15               17.0  
68172                    2018-02-08               17.0

In [55]:
mean_fact_delivery_dur = orders_df["fact_delivery_dur"].mean()
print(f"Fact Delivery Duration Mean: {mean_fact_delivery_dur} days")

orders_df['order_delivered_customer_date'].fillna(
    value=orders_df['order_estimated_delivery_date'] -
    pd.to_timedelta(mean_fact_delivery_dur, unit='m'),
    inplace=True
)

orders_df.drop("fact_delivery_dur", axis=1, inplace=True)

Fact Delivery Duration Mean: 11.00884166882612 days


In [56]:
orders_df.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

In [58]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 97658 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       97658 non-null  object        
 1   customer_id                    97658 non-null  object        
 2   order_status                   97658 non-null  object        
 3   order_purchase_timestamp       97658 non-null  datetime64[ns]
 4   order_approved_at              97658 non-null  datetime64[ns]
 5   order_delivered_carrier_date   97658 non-null  datetime64[ns]
 6   order_delivered_customer_date  97658 non-null  datetime64[ns]
 7   order_estimated_delivery_date  97658 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.7+ MB


### Cleaning data `products_df`

In [59]:
products_df.dropna(inplace=True)
products_df.isna().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

In [60]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32340 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32340 non-null  object 
 1   product_category_name       32340 non-null  object 
 2   product_name_lenght         32340 non-null  float64
 3   product_description_lenght  32340 non-null  float64
 4   product_photos_qty          32340 non-null  float64
 5   product_weight_g            32340 non-null  float64
 6   product_length_cm           32340 non-null  float64
 7   product_height_cm           32340 non-null  float64
 8   product_width_cm            32340 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.5+ MB


# Exploratory Data Analysis (EDA) Session
---

## Business Question Definition

1. Kapan terakhir kali pelanggan melakukan pembelian?
2. Seberapa sering pelanggan melakukan pembelian?
3. Berapa banyak uang yang dihabiskan pelanggan?
4. 

## `customers_df` Exploration

In [61]:
customers_df.describe(include="all")

customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        customer_zip_code_prefix customer_city customer_state  
count               99441.000000         99441          99441  
unique                       NaN          4119             27  
top                          NaN     sao paulo             SP  
freq                         NaN         15540          41746  
mean                35137.474583           NaN            NaN  
std                 29797.938996           NaN            NaN  
min                  1003.000000           NaN            NaN  
25%                 11347.000000           NaN            NaN  
50%                 24416.000000           NaN            NaN  
75%                 58900.000000           NaN            NaN  
max                 99990.000000           NaN            NaN

In [64]:
customers_df.groupby(by="customer_city").customer_id.nunique().sort_values(ascending=False)

customer_city
sao paulo         15540
rio de janeiro     6882
belo horizonte     2773
brasilia           2131
curitiba           1521
                  ...  
ibiara                1
rio espera            1
rio dos indios        1
rio dos cedros        1
lagoao                1
Name: customer_id, Length: 4119, dtype: int64

In [65]:
customers_df.groupby(by="customer_state").customer_id.nunique().sort_values(ascending=False)

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: customer_id, dtype: int64

In [66]:
customers_df.groupby(by="customer_zip_code_prefix").customer_id.nunique(
).sort_values(ascending=False)

customer_zip_code_prefix
22790    142
24220    124
22793    121
24230    117
22775    110
        ... 
58421      1
58411      1
58398      1
58396      1
99990      1
Name: customer_id, Length: 14994, dtype: int64

## `geolocation_df` Exploration

In [67]:
geolocation_df.describe(include="all")

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
count                 738332.000000    738332.000000    738332.000000   
unique                          NaN              NaN              NaN   
top                             NaN              NaN              NaN   
freq                            NaN              NaN              NaN   
mean                   38316.086800       -20.998353       -46.461098   
std                    30632.496675         5.892315         4.393705   
min                     1001.000000       -36.605374      -101.466766   
25%                    12600.000000       -23.603061       -48.867822   
50%                    29144.000000       -22.873588       -46.647278   
75%                    65950.000000       -19.923336       -43.836974   
max                    99990.000000        45.065933       121.105394   

       geolocation_city geolocation_state  
count            738332            738332  
unique             8011                27  
top           sao paulo                SP  
freq              79927            285976  
mean                NaN               NaN  
std                 NaN               NaN  
min                 NaN               NaN  
25%                 NaN               NaN  
50%                 NaN               NaN  
75%                 NaN               NaN  
max                 NaN               NaN